File: src/render_x11.c
// SPDX-License-Identifier: MIT
/*
 * render_x11.c - X11 render backend.
 *
 * Creates an override-redirect, always-on-top window positioned at the same
 * panel rectangle used by the framebuffer backend.  The window is unmapped
 * when no overlay is shown, so it is invisible and does not interfere with
 * other applications.
 *
 * Glyph rendering reuses the same font8x16 bitmap data as the framebuffer
 * backend (via XFillRectangle), so both backends produce identical-looking
 * output.
 *
 * Color allocation uses XAllocColor so the code is correct for all X visual
 * types (PseudoColor, TrueColor, …).
 *
 * Only compiled when WITH_X11 is defined (set by the Makefile).
 */

#ifdef WITH_X11

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>

#include "font8x16.h"
#include "keyoverlay.h"
#include "render_x11.h"

/* ------------------------------------------------------------------ */
/* Internal state                                                      */
/* ------------------------------------------------------------------ */

struct x11_backend {
	render_backend  base;   /* MUST be first */
	Display        *dpy;
	int             screen;
	Window          win;
	GC              gc;
	struct panel    panel;
	bool            shown;

	/* Pre-allocated pixel values for each named color. */
	unsigned long   px_panel;
	unsigned long   px_border;
	unsigned long   px_cell;
	unsigned long   px_cell_empty;
	unsigned long   px_text;
	unsigned long   px_title;
};

/* ------------------------------------------------------------------ */
/* Color helpers                                                       */
/* ------------------------------------------------------------------ */

static unsigned long alloc_color(Display *dpy, int screen, struct rgb c)
{
	XColor xc;
	xc.red   = (unsigned short)c.r << 8;
	xc.green = (unsigned short)c.g << 8;
	xc.blue  = (unsigned short)c.b << 8;
	xc.flags = DoRed | DoGreen | DoBlue;
	if (XAllocColor(dpy, DefaultColormap(dpy, screen), &xc))
		return xc.pixel;
	/* Fallback: compose a 24-bit value for TrueColor displays. */
	return ((unsigned long)c.r << 16) |
	       ((unsigned long)c.g <<  8) |
	        (unsigned long)c.b;
}

/* ------------------------------------------------------------------ */
/* Primitive drawing (onto the overlay window)                         */
/* ------------------------------------------------------------------ */

static void x11_fill_rect(struct x11_backend *b, int x, int y, int w, int h,
			   unsigned long px)
{
	XSetForeground(b->dpy, b->gc, px);
	XFillRectangle(b->dpy, b->win, b->gc, x, y,
		       (unsigned)w, (unsigned)h);
}

static void x11_draw_glyph(struct x11_backend *b, int x, int y, char ch,
			    unsigned long fg)
{
	if ((unsigned char)ch < FONT_FIRST || (unsigned char)ch > FONT_LAST)
		ch = '?';
	const unsigned char *g =
		&font8x16[((unsigned char)ch - FONT_FIRST) * FONT_H];
	XSetForeground(b->dpy, b->gc, fg);
	for (int row = 0; row < FONT_H; row++) {
		unsigned char bits = g[row];
		for (int col = 0; col < FONT_W; col++) {
			if (!(bits & (0x80 >> col)))
				continue;
			XFillRectangle(b->dpy, b->win, b->gc,
				       x + col * FONT_SCALE,
				       y + row * FONT_SCALE,
				       FONT_SCALE, FONT_SCALE);
		}
	}
}

static void x11_draw_text(struct x11_backend *b, int x, int y, const char *s,
			   unsigned long fg)
{
	for (; *s; s++, x += GLYPH_W)
		x11_draw_glyph(b, x, y, *s, fg);
}

/* ------------------------------------------------------------------ */
/* Layout rendering                                                    */
/* ------------------------------------------------------------------ */

static void draw_layout_x11(struct x11_backend *b, const Layout *L)
{
	/*
	 * The window covers exactly the panel rectangle, so all coordinates
	 * below are relative to the window's top-left corner (i.e. subtract
	 * panel.x / panel.y from the absolute positions used by the fb backend).
	 */
	int pw = b->panel.w;
	int ph = b->panel.h;

	x11_fill_rect(b, 0, 0, pw, ph, b->px_panel);

	/* Double border */
	XSetForeground(b->dpy, b->gc, b->px_border);
	XDrawRectangle(b->dpy, b->win, b->gc, 0, 0,
		       (unsigned)(pw - 1), (unsigned)(ph - 1));
	XDrawRectangle(b->dpy, b->win, b->gc, 1, 1,
		       (unsigned)(pw - 3), (unsigned)(ph - 3));

	int x0 = PANEL_PAD;
	int y  = PANEL_PAD;

	x11_draw_text(b, x0, y, L->title, b->px_title);
	y += GLYPH_H + TITLE_GAP;

	for (int ri = 0; ri < L->nrows; ri++) {
		const Row *r = &L->rows[ri];
		int x = x0 + (r->indent_half * MIN_CELL_W) / 2;
		for (int ci = 0; ci < r->n; ci++) {
			const char *label = r->label[ci];
			int   cw    = cell_width(label);
			bool  empty = (label[0] == '\0');
			x11_fill_rect(b, x, y, cw, CELL_H,
				      empty ? b->px_cell_empty : b->px_cell);
			if (!empty) {
				int tw = (int)strlen(label) * GLYPH_W;
				int tx = x + (cw - tw) / 2;
				int ty = y + (CELL_H - GLYPH_H) / 2;
				x11_draw_text(b, tx, ty, label, b->px_text);
			}
			x += cw + HGAP;
		}
		y += CELL_H + VGAP;
	}

	XFlush(b->dpy);
}

/* ------------------------------------------------------------------ */
/* Backend vtable                                                      */
/* ------------------------------------------------------------------ */

static void x11_show(render_backend *base, const Layout *L)
{
	struct x11_backend *b = (struct x11_backend *)base;
	if (!b->shown) {
		XMapRaised(b->dpy, b->win);
		b->shown = true;
	}
	draw_layout_x11(b, L);
}

static void x11_hide(render_backend *base)
{
	struct x11_backend *b = (struct x11_backend *)base;
	if (!b->shown)
		return;
	XUnmapWindow(b->dpy, b->win);
	XFlush(b->dpy);
	b->shown = false;
}

static void x11_close(render_backend *base)
{
	struct x11_backend *b = (struct x11_backend *)base;
	if (b->gc)
		XFreeGC(b->dpy, b->gc);
	if (b->win)
		XDestroyWindow(b->dpy, b->win);
	if (b->dpy)
		XCloseDisplay(b->dpy);
	free(b);
}

/* ------------------------------------------------------------------ */
/* Constructor                                                         */
/* ------------------------------------------------------------------ */

render_backend *render_x11_create(int verbose)
{
	struct x11_backend *b = calloc(1, sizeof(*b));
	if (!b)
		return NULL;

	b->base.show  = x11_show;
	b->base.hide  = x11_hide;
	b->base.close = x11_close;

	b->dpy = XOpenDisplay(NULL);
	if (!b->dpy) {
		fprintf(stderr, "keyoverlay: cannot open X display\n");
		goto err_free;
	}
	b->screen = DefaultScreen(b->dpy);

	int screen_w = DisplayWidth(b->dpy,  b->screen);
	int screen_h = DisplayHeight(b->dpy, b->screen);

	if (verbose)
		fprintf(stderr, "keyoverlay: X11 display %dx%d\n",
			screen_w, screen_h);

	b->panel = compute_panel(screen_w, screen_h);

	/* Pre-allocate colors. */
	b->px_panel      = alloc_color(b->dpy, b->screen, COL_PANEL);
	b->px_border     = alloc_color(b->dpy, b->screen, COL_BORDER);
	b->px_cell       = alloc_color(b->dpy, b->screen, COL_CELL);
	b->px_cell_empty = alloc_color(b->dpy, b->screen, COL_CELL_EMPTY);
	b->px_text       = alloc_color(b->dpy, b->screen, COL_TEXT);
	b->px_title      = alloc_color(b->dpy, b->screen, COL_TITLE);

	/* Create an override-redirect window (bypasses the window manager). */
	XSetWindowAttributes attr = {0};
	attr.override_redirect = True;
	attr.background_pixel  = b->px_panel;

	b->win = XCreateWindow(
		b->dpy,
		RootWindow(b->dpy, b->screen),
		b->panel.x, b->panel.y,
		(unsigned)b->panel.w, (unsigned)b->panel.h,
		0,                          /* border width */
		DefaultDepth(b->dpy, b->screen),
		InputOutput,
		DefaultVisual(b->dpy, b->screen),
		CWOverrideRedirect | CWBackPixel,
		&attr);

	/* Hint to EWMH compositors that this is a notification-style window. */
	Atom wm_type      = XInternAtom(b->dpy, "_NET_WM_WINDOW_TYPE",        False);
	Atom type_notif   = XInternAtom(b->dpy, "_NET_WM_WINDOW_TYPE_NOTIFICATION", False);
	XChangeProperty(b->dpy, b->win, wm_type, XA_ATOM, 32,
			PropModeReplace, (unsigned char *)&type_notif, 1);

	/* Set a descriptive title for debugging. */
	XStoreName(b->dpy, b->win, "keyoverlay");

	b->gc = XCreateGC(b->dpy, b->win, 0, NULL);
	if (!b->gc) {
		fprintf(stderr, "keyoverlay: XCreateGC failed\n");
		goto err_destroy_win;
	}

	/* Window starts unmapped (invisible) until show() is called. */
	XFlush(b->dpy);
	return &b->base;

err_destroy_win:
	XDestroyWindow(b->dpy, b->win);
	XCloseDisplay(b->dpy);
err_free:
	free(b);
	return NULL;
}

#endif /* WITH_X11 */

File: src/render_x11.zig
// SPDX-License-Identifier: MIT
//
// render_x11.zig - X11 render backend.
//
// Creates an override-redirect, always-on-top window positioned at the same
// panel rectangle used by the framebuffer backend.  The window is unmapped
// when no overlay is shown, so it is invisible and does not interfere with
// other applications.  Glyph rendering reuses the same font8x16 bitmap data
// (via XFillRectangle), so both backends produce identical-looking output.
//
// This file is only referenced (and therefore only compiled, pulling in the
// libX11 dependency) when the program is built with -Dx11=true.

const std = @import("std");
const font = @import("font8x16.zig");
const lo = @import("layout.zig");

const Layout = lo.Layout;
const Panel = lo.Panel;
const Rgb = lo.Rgb;
const Backend = @import("render.zig").Backend;

const x = @cImport({
    @cInclude("X11/Xlib.h");
    @cInclude("X11/Xutil.h");
    @cInclude("X11/Xatom.h");
});

pub const X11Backend = struct {
    alloc: std.mem.Allocator,
    dpy: *x.Display,
    screen: c_int,
    win: x.Window,
    gc: x.GC,
    panel: Panel,
    shown: bool,

    px_panel: c_ulong,
    px_border: c_ulong,
    px_cell: c_ulong,
    px_cell_empty: c_ulong,
    px_text: c_ulong,
    px_title: c_ulong,

    const vtable = Backend.VTable{
        .show = showThunk,
        .hide = hideThunk,
        .deinit = deinitThunk,
    };

    pub fn backend(self: *X11Backend) Backend {
        return .{ .ptr = self, .vtable = &vtable };
    }

    // -------------------------------------------------------------- //
    // Primitive drawing (onto the overlay window)                     //
    // -------------------------------------------------------------- //

    fn fillRect(self: *X11Backend, px_x: i32, px_y: i32, w: i32, h: i32, px: c_ulong) void {
        _ = x.XSetForeground(self.dpy, self.gc, px);
        _ = x.XFillRectangle(self.dpy, self.win, self.gc, px_x, px_y, @intCast(w), @intCast(h));
    }

    fn drawGlyph(self: *X11Backend, px_x: i32, px_y: i32, ch: u8, fg: c_ulong) void {
        var cc = ch;
        if (cc < font.FONT_FIRST or cc > font.FONT_LAST) cc = '?';
        const base = (@as(usize, cc) - font.FONT_FIRST) * font.FONT_H;
        _ = x.XSetForeground(self.dpy, self.gc, fg);
        var row: usize = 0;
        while (row < font.FONT_H) : (row += 1) {
            const bits = font.font8x16[base + row];
            var col: usize = 0;
            while (col < font.FONT_W) : (col += 1) {
                if (bits & (@as(u8, 0x80) >> @intCast(col)) == 0) continue;
                _ = x.XFillRectangle(
                    self.dpy,
                    self.win,
                    self.gc,
                    px_x + @as(i32, @intCast(col)) * lo.FONT_SCALE,
                    px_y + @as(i32, @intCast(row)) * lo.FONT_SCALE,
                    lo.FONT_SCALE,
                    lo.FONT_SCALE,
                );
            }
        }
    }

    fn drawText(self: *X11Backend, px_x: i32, px_y: i32, s: []const u8, fg: c_ulong) void {
        var cx = px_x;
        for (s) |ch| {
            self.drawGlyph(cx, px_y, ch, fg);
            cx += lo.GLYPH_W;
        }
    }

    // -------------------------------------------------------------- //
    // Layout rendering                                                //
    // -------------------------------------------------------------- //

    fn drawLayout(self: *X11Backend, l: *const Layout) void {
        // The window covers exactly the panel rectangle, so all coordinates
        // here are relative to the window's top-left corner.
        const pw = self.panel.w;
        const ph = self.panel.h;

        self.fillRect(0, 0, pw, ph, self.px_panel);

        // Double border.
        _ = x.XSetForeground(self.dpy, self.gc, self.px_border);
        _ = x.XDrawRectangle(self.dpy, self.win, self.gc, 0, 0, @intCast(pw - 1), @intCast(ph - 1));
        _ = x.XDrawRectangle(self.dpy, self.win, self.gc, 1, 1, @intCast(pw - 3), @intCast(ph - 3));

        const x0 = lo.PANEL_PAD;
        var y: i32 = lo.PANEL_PAD;

        self.drawText(x0, y, l.title, self.px_title);
        y += lo.GLYPH_H + lo.TITLE_GAP;

        for (l.rows) |*r| {
            var cx = x0 + @divTrunc(r.indent_half * lo.MIN_CELL_W, 2);
            for (r.label) |label| {
                const cw = lo.cellWidth(label);
                const empty = label.len == 0;
                self.fillRect(cx, y, cw, lo.CELL_H, if (empty) self.px_cell_empty else self.px_cell);
                if (!empty) {
                    const tw = @as(i32, @intCast(label.len)) * lo.GLYPH_W;
                    const tx = cx + @divTrunc(cw - tw, 2);
                    const ty = y + @divTrunc(lo.CELL_H - lo.GLYPH_H, 2);
                    self.drawText(tx, ty, label, self.px_text);
                }
                cx += cw + lo.HGAP;
            }
            y += lo.CELL_H + lo.VGAP;
        }

        _ = x.XFlush(self.dpy);
    }

    // -------------------------------------------------------------- //
    // Backend vtable                                                  //
    // -------------------------------------------------------------- //

    fn show(self: *X11Backend, l: *const Layout) void {
        if (!self.shown) {
            _ = x.XMapRaised(self.dpy, self.win);
            self.shown = true;
        }
        self.drawLayout(l);
    }

    fn hide(self: *X11Backend) void {
        if (!self.shown) return;
        _ = x.XUnmapWindow(self.dpy, self.win);
        _ = x.XFlush(self.dpy);
        self.shown = false;
    }

    fn deinit(self: *X11Backend) void {
        const alloc = self.alloc;
        if (self.gc != null) _ = x.XFreeGC(self.dpy, self.gc);
        if (self.win != 0) _ = x.XDestroyWindow(self.dpy, self.win);
        _ = x.XCloseDisplay(self.dpy);
        alloc.destroy(self);
    }

    fn showThunk(ptr: *anyopaque, l: *const Layout) void {
        show(@ptrCast(@alignCast(ptr)), l);
    }
    fn hideThunk(ptr: *anyopaque) void {
        hide(@ptrCast(@alignCast(ptr)));
    }
    fn deinitThunk(ptr: *anyopaque) void {
        deinit(@ptrCast(@alignCast(ptr)));
    }
};

// ------------------------------------------------------------------ //
// Color helper                                                        //
// ------------------------------------------------------------------ //

fn allocColor(dpy: *x.Display, screen: c_int, col: Rgb) c_ulong {
    var xc: x.XColor = std.mem.zeroes(x.XColor);
    xc.red = @as(c_ushort, col.r) << 8;
    xc.green = @as(c_ushort, col.g) << 8;
    xc.blue = @as(c_ushort, col.b) << 8;
    xc.flags = @intCast(x.DoRed | x.DoGreen | x.DoBlue);
    if (x.XAllocColor(dpy, x.XDefaultColormap(dpy, screen), &xc) != 0)
        return xc.pixel;
    // Fallback: compose a 24-bit value for TrueColor displays.
    return (@as(c_ulong, col.r) << 16) | (@as(c_ulong, col.g) << 8) | col.b;
}

// ------------------------------------------------------------------ //
// Constructor                                                         //
// ------------------------------------------------------------------ //

pub fn create(alloc: std.mem.Allocator, verbose: bool) ?Backend {
    const self = alloc.create(X11Backend) catch return null;
    errdefer alloc.destroy(self);

    const dpy = x.XOpenDisplay(null) orelse {
        std.debug.print("keyoverlay: cannot open X display\n", .{});
        return null;
    };
    errdefer _ = x.XCloseDisplay(dpy);

    const screen = x.XDefaultScreen(dpy);
    const screen_w = x.XDisplayWidth(dpy, screen);
    const screen_h = x.XDisplayHeight(dpy, screen);

    if (verbose)
        std.debug.print("keyoverlay: X11 display {d}x{d}\n", .{ screen_w, screen_h });

    const panel = lo.computePanel(screen_w, screen_h);

    const px_panel = allocColor(dpy, screen, lo.COL_PANEL);

    // Create an override-redirect window (bypasses the window manager).
    var attr: x.XSetWindowAttributes = std.mem.zeroes(x.XSetWindowAttributes);
    attr.override_redirect = 1; // True
    attr.background_pixel = px_panel;

    const win = x.XCreateWindow(
        dpy,
        x.XRootWindow(dpy, screen),
        panel.x,
        panel.y,
        @intCast(panel.w),
        @intCast(panel.h),
        0, // border width
        x.XDefaultDepth(dpy, screen),
        x.InputOutput,
        x.XDefaultVisual(dpy, screen),
        x.CWOverrideRedirect | x.CWBackPixel,
        &attr,
    );
    errdefer _ = x.XDestroyWindow(dpy, win);

    // Hint to EWMH compositors that this is a notification-style window.
    const wm_type = x.XInternAtom(dpy, "_NET_WM_WINDOW_TYPE", x.False);
    var type_notif = x.XInternAtom(dpy, "_NET_WM_WINDOW_TYPE_NOTIFICATION", x.False);
    _ = x.XChangeProperty(dpy, win, wm_type, x.XA_ATOM, 32, x.PropModeReplace, @ptrCast(&type_notif), 1);

    // Set a descriptive title for debugging.
    _ = x.XStoreName(dpy, win, "keyoverlay");

    const gc = x.XCreateGC(dpy, win, 0, null);
    if (gc == null) {
        std.debug.print("keyoverlay: XCreateGC failed\n", .{});
        return null;
    }

    self.* = .{
        .alloc = alloc,
        .dpy = dpy,
        .screen = screen,
        .win = win,
        .gc = gc,
        .panel = panel,
        .shown = false,
        .px_panel = px_panel,
        .px_border = allocColor(dpy, screen, lo.COL_BORDER),
        .px_cell = allocColor(dpy, screen, lo.COL_CELL),
        .px_cell_empty = allocColor(dpy, screen, lo.COL_CELL_EMPTY),
        .px_text = allocColor(dpy, screen, lo.COL_TEXT),
        .px_title = allocColor(dpy, screen, lo.COL_TITLE),
    };

    // Window starts unmapped (invisible) until show() is called.
    _ = x.XFlush(dpy);
    return self.backend();
}
